### PR TITLE
Fix `calculate_pre_verification_gas` not to happen overflow problem

### PR DESCRIPTION
**Pull request by winor30**
### Description

The previous implementation had a potential for overflow. The logic has been modified to prevent this issue.

### Changes

- Removed the use of floating point calculations which might lead to inaccuracies and potential overflows.
- Replaced the previous calculation logic with a more straightforward and safer approach using `saturating_add`, `checked_div`, and other safe arithmetic operations.
- Added tests to ensure the new logic works correctly, especially in scenarios with large user operations and potential overflow situations.

### Relate
`1. type coversion` in https://github.com/Vid201/silius/issues/191

### CODE REVIEW

File: crates/uopool/src/utils.rs [hunk 1]
```diff
@@ -62,25 +62,36 @@ impl Overhead {
...
+        let word_cost = self
+            .per_user_op_word
+            .saturating_mul(U256::from(word_len))
+            .checked_div(U256::from(32))
+            .unwrap_or_default()
+            .saturating_add(U256::from(if (word_len) % 32 > 0 { 1 } else { 0 }));
+
+        let fixed_divided_by_bundle_size = self
+            .fixed
+            .saturating_mul(U256::from(1_000_000))
```
Comment (zsluedem) [inline crates/uopool/src/utils.rs:85]: I am curious why 1_000_000 here.

File: crates/uopool/src/utils.rs [hunk 2]
```diff
@@ -62,25 +62,36 @@ impl Overhead {
...
+            acc.saturating_add(*byte_cost)
+        });
+
+        let word_len = uo_pack.len() + 31;
+        let word_cost = self
+            .per_user_op_word
+            .saturating_mul(U256::from(word_len))
+            .checked_div(U256::from(32))
+            .unwrap_or_default()
+            .saturating_add(U256::from(if (word_len) % 32 > 0 { 1 } else { 0 }));
```
Comment (zsluedem) [inline crates/uopool/src/utils.rs:81]: Why is this needed? You want to get the same result from old calculation, right?

File: crates/uopool/src/utils.rs [hunk 3]
```diff
@@ -146,6 +157,91 @@ pub mod tests {
         assert_eq!(gas_oh.calculate_pre_verification_gas(&uo), 45340.into());
     }
 
+    #[test]
+    fn pre_verification_gas_calculation_with_large_user_operation() {
```
Comment (zsluedem) [inline crates/uopool/src/utils.rs:161]: love it!

File: crates/uopool/src/utils.rs [hunk 4]
```diff
@@ -146,6 +157,91 @@ pub mod tests {
...
+            sender: "0xAB7e2cbFcFb6A5F33A75aD745C3E5fB48d689B54"
+                .parse()
+                .unwrap(),
+            nonce: U256::max_value(),
+            init_code: Bytes::from(vec![255; 1024]), // Large init_code
+            call_data: Bytes::from(vec![255; 1024]), // Large call_data
+            call_gas_limit: U256::max_value(),
+            verification_gas_limit: U256::max_value(),
+            pre_verification_gas: U256::max_value(),
+            max_fee_per_gas: U256::max_value(),
```
Comment (zsluedem) [inline crates/uopool/src/utils.rs:203]: It could be divided by zero if `max_fee_per_gas` is zero based on #191 . Would you mind also fixing it?

**Review by zsluedem — COMMENTED** [review]
Very awesome work~! I really love your pr and want to merge it though I had some questions. 

https://github.com/Vid201/silius/blob/33dc6e686fd2fff4b26e0c5ceadce20e3cf5e739/crates/uopool/src/utils.rs#L78-L82
Do you think we overflow could also be possible here? Would you mind addressing it?

It is ok to refuse if you don't want to.

Thanks very much~!!

File: crates/uopool/src/utils.rs [hunk 5]
```diff
@@ -94,9 +108,39 @@ impl Overhead {
...
+/// performs the division, and rounds up if there is a remainder.
+///
+/// # Examples
+///
+/// ```
+/// # use your_crate::U256; // Replace with the actual import
+/// let result = div_ceil(U256::from(10), U256::from(3));
+/// assert_eq!(result, U256::from(4));
+/// ```
+pub fn div_ceil(numerator: U256, denominator: U256) -> U256 {
```
Comment (winor30) [inline crates/uopool/src/utils.rs:132]:
```suggestion
fn div_ceil(numerator: U256, denominator: U256) -> U256 {
```

File: crates/uopool/src/utils.rs (hunk 1 above)
Comment (winor30) [reply crates/uopool/src/utils.rs]: This was a scaling constant to prevent rounding of the division result.
However, 1_000_000 could round the quotient, so [`div_ceil`](https://github.com/Vid201/silius/pull/214/files#diff-4edf5911f99a49053c851d9441d21d3054641602def24869489b642956c5479bR132) was provided.
div_ceil is also used in other parts of the division.

File: crates/uopool/src/utils.rs (hunk 2 above)
Comment (winor30) [reply crates/uopool/src/utils.rs]: This process rounds up the decimal point when the result returns a decimal point.
If the result of `% 32` is non-zero, the decimal point is present and 1 is added.

But this has also been changed to use [div_ceil](https://github.com/Vid201/silius/pull/214/files#diff-4edf5911f99a49053c851d9441d21d3054641602def24869489b642956c5479bR122-R123).

File: crates/uopool/src/utils.rs (hunk 4 above)
Comment (winor30) [reply crates/uopool/src/utils.rs:236]: ok! i will fix it

Comment (winor30) [reply crates/uopool/src/utils.rs:240]: Is it correct like this?
[`6e3b6cd` (#214)](https://github.com/Vid201/silius/pull/214/commits/6e3b6cde063f5a72c3d3ed6cbbc51f74f35eb3f8)

**Review by zsluedem — APPROVED** [review]
Very good work~! Thanks for addressing all the comments and the additional issues~! 

@Vid201 do you think this is good?

**Comment by Vid201** [comment]
> Very good work~! Thanks for addressing all the comments and the additional issues~!
> 
> @Vid201 do you think this is good?

Looks good to me! GJ @winor30 

Can we just squash all commits when merging @zsluedem

**Comment by zsluedem** [comment]
@winor30 Please squash the commits into one. Then I will merge it. Thanks very much~!

**Comment by winor30** [comment]
@Vid201 @zsluedem thank you for your review!

@zsluedem 
> Please squash the commits into one. Then I will merge it. Thanks very much~!

I could squash the commits myself, but I thought it would be better to have a squash merge.
like below
image

**Comment by zsluedem** [comment]
@winor30 Thanks. I didn't know github got this feature actually. Anyway, really appreciate your help.
